keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1061,28 +1062,91 @@ Details to consider:
+it was copy-pasted. A variable, on the other hand, will only need to be updated
+in one place.
+
+###### Reusing/memoizing an expensive computation
+
+For a CEL expression that runs in O(n^2) time or worse (or otherwise
+takes a significant amount of time to execute), it would be nice to only run
+it when necessary. For instance, if multiple validation expressions used the
+same expensive expression, that expression could be refactored out into a
+variable.
+
 ##### Variables
 
-Each CEL "program" is a single expression. There is no support for vaiable
+Each CEL "program" is a single expression. There is no support for variable
 assignment. This can result in redundant code to traverse maps/arrays or
 dereference particular fields.
 
 We can support this in much the same way as cel-policy-template `terms`. These
 can be lazily evaluated while the validation expressions are evaluated
-(cel-policy-template does this). The results can also be memoized to avoid
-repeated evaluations if they are shared across validations.
+(cel-policy-template does this).
+
+A policy can include an additional `variables` section. This is an array
+containing one or more `name` and `expression` pairs, which can be used/re-used by
+the policy's validation expressions. These results are memoized on a
+per-validation basis, so if multiple expressions use the same spec variables,
+the expression that calculates the variable's value will only be ran once.
+
+The variables can be accessed as members of `variables`, which is an object
+that is exposed to CEL expressions (both validation expressions as well as
+other variables).
+
+For example:
 
 ```yaml
   variables:
     - name: metadataList
       expression: "spec.list.map(x, x.metadata)"
     - name: itemMetadataNames
-      expression: "metadataList.map(m, m.name)"
+      expression: "variables.metadataList.map(m, m.name)"
   validations:
-    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
-    - expression: "itemMetadataNames.exists(name, name == 'required')"
+    - expression: "variables.itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "variables.itemMetadataNames.exists(name, name == 'required')"
 ```
 
+Variable names must be valid CEL names, and not already used by Kubernetes
+in some way (i.e, `request` is not considered valid). What constitutes a
+valid CEL name can be found at CEL's [language definition](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax) under `IDENT`.
+This validity is checked at creation time.
+
+For runtime cost limit purposes, variables count towards the runtime cost limit
+once per policy. The cost of each variable is computed when it is first evaluated in an
+expression, mirroring how the cost limit would be calculated if the variable's
+expression was embedded verbatim.  If the runtime cost limit is exceeded in the
+process, then evaluation halts. Whether or not the request actually fails
```

Comment:
When the cost limit is hit, how will it be attributed? Imagine there is a variable with a high cost and multiple validation rules all with near zero costs, will the cost limit error make it clear that the variable is responsible for the limit being hit?  I'm curious about this both for the per-policy limit and per-expression limit.

cc @cici37

Reply:
Feels like a strange situation where a policy has to be "primed" a few times before it starts working. If there are N high-cost variables, and each execution has budget for only one variable, then the policy may need N execution to warm up.

Reply:
> When the cost limit is hit, how will it be attributed? Imagine there is a variable with a high cost and multiple validation rules all with near zero costs, will the cost limit error make it clear that the variable is responsible for the limit being hit? I'm curious about this both for the per-policy limit and per-expression limit.

I think that noting execution failed during variable evaluation is good. Maybe if we track how much time we spend during variable evaluation vs. validation rule evaluation we could provide a cost breakdown in the error message - x% time spent in variable vs. y% time spent with rule?

> Feels like a strange situation where a policy has to be "primed" a few times before it starts working. If there are N high-cost variables, and each execution has budget for only one variable, then the policy may need N execution to warm up.

I'm not sure that'd work, since expressions are only memoized on a per-validation basis, so subsequent executions would be starting from scratch.

Reply:
I missed the "per-validation" part.

Reply:
(resolved)

Reply:
My $.02:

- I think any per-expression runtime limit should apply to each variable expression. If the expression hits the limit when being evaluated (regardless of the fact that it is being lazy-evaluated to satisfy some other variable or expression), the runtime limit error should be attributed to that variable. For expressions and variables that depend on other variables, we should not include the runtime costs of evaluating the dependents.
- The per-policy limit should probably be phrased to not assign blame to a particular variable or expression unless we can summarize ALL costs, which I'm not convinced is worth the investment (we don't do that today in CRD validation do we?)

Reply:
> I think any per-expression runtime limit should apply to each variable expression. If the expression hits the limit when being evaluated (regardless of the fact that it is being lazy-evaluated to satisfy some other variable or expression), the runtime limit error should be attributed to that variable. For expressions and variables that depend on other variables, we should not include the runtime costs of evaluating the dependents.

Gotcha, I'll make it clear that the per-expression limit still applies to variables, and that the original expression will be listed as the cause for per-policy failures.

> The per-policy limit should probably be phrased to not assign blame to a particular variable or expression unless we can summarize ALL costs, which I'm not convinced is worth the investment (we don't do that today in CRD validation do we?)

Right, we don't do this with CRDs, so we can keep things consistent and not do it here either.